KDTree Data Structure for Center and Corner Nodes

aaronzedwick opened the pull request:
<!--  The PR title should summarize the changes, for example "Add `Grid._build_face_dimension` function".
      Avoid non-descriptive titles such as "Addresses issue #229". -->

<!--  Replace XXX with the number of the issue that this PR will resolve. If this PR closed more than one,
      you may add a comma separated sequence-->
Closes #496

## Overview
<!--  Please provide a few bullet points summarizing the changes in this PR. This should include
      points on any bug fixes, new functions, or other changes that have been made. -->
Adds a KDTree data structure for center nodes (`Mesh2_face_cart_x`, `Mesh2_face_cart_y`, `Mesh2_face_cart_z`) and corner nodes (`Mesh2_node_cart_x`, `Mesh2_node_cart_y`, `Mesh2_node_cart_z`)  for grid files. This can be used for nearest neighbor queries as well as querying in a specific radius. This implementation mirrors the BallTree data structure #406 and uses `sklearn.neighbors.KDTree` with a minkowski distance for querying.  

## Expected Usage
<!--  If you are adding a feature into the Internal API, please produce a short example of it in action.
      You may ignore this step if it is not applicable (comment out this section). -->
```Python
import uxarray as ux

grid_path = "/path/to/grid.nc"

uxgrid = ux.open_grid(grid_path)

# find the nearest neighbor to (0.0, 0.0, 1.0) using corner nodes
d, ind = uxgrid.get_kd_tree(tree_type="nodes").query([0.0, 0.0, 1.0])

# find the 5 nearest neighbors to (0.0, 0.0, 1.0) using face centers
d, ind = uxgrid.get_kd_tree(tree_type="face centers").query([1.0, 0.0, 0.0], k=5)

# find all face centers within a radius of 80.00 from (0.0, 0.0, 1.0)
d, ind = uxgrid.get_kd_tree(tree_type="face centers").query_radius([0.0, 0.0, 1.0], r=80.00)
```

## PR Checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. If an entire section doesn't
apply to this PR, comment it out or delete it. -->

**General**
- [x] An issue is linked created and linked
- [x] Add appropriate labels
- [x] Filled out Overview and Expected Usage (if applicable) sections

**Testing**
- [x] Adequate tests are created if there is new functionality
- [x] Tests cover all possible logical paths in your function
- [x] Tests are not too basic (such as simply calling a function and nothing else)

**Documentation**
- [x] Docstrings have been added to all new functions
- [x] Docstrings have updated with any function changes
- [x] Internal functions have a preceding underscore (`_`) and have been added to `docs/internal_api/index.rst`
- [x] User functions have been added to `docs/user_api/index.rst`

**Examples**
- [x] Any new notebook examples added to `docs/examples/` folder
- [x] Clear the output of all cells before committing
- [x] New notebook files added to `docs/examples.rst` toctree
- [x] New notebook files added to new entry in `docs/gallery.yml` with appropriate thumbnail photo in `docs/_static/thumbnails/`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

**PR Etiquette Reminders**
- This PR should be listed as a draft PR until you are ready to request reviewers

- After making changes in accordance with the reviews, re-request your reviewers

- Do *not* mark conversations as resolved if you didn't start them

- Do mark conversations as resolved *if you opened them* and are satisfied with the changes/discussion.
-->

philipc2 commented on uxarray/grid/neighbors.py:
```diff
@@ -268,3 +493,25 @@ def _prepare_xy_for_query(xy, use_radians):
     return xy
+
+
+def _prepare_cart_xyz_for_query(xyz):
```
```suggestion
def _prepare_xyz_for_query(xyz):
```

this name should suffice

philipc2 commented on uxarray/grid/neighbors.py:
```diff
@@ -3,11 +3,236 @@
+            raise AssertionError(
+                f"The value of r must be greater than or equal to zero.")
+
+        r = np.deg2rad(r)
```
I'd be careful with this.

The reason we did the `deg2rad` conversion is because our `BallTree` used the haversine distance, which works in radians.

Here we are using the mikowski distance in Cartesian space, I'm not sure what the physical meaning of our "r" is. Can you look into it?

aaronzedwick replied:
Yeah, I think you are right, it does make sense that r should be in whatever measurement you have for your input grid.

aaronzedwick commented on uxarray/grid/neighbors.py:
```diff
@@ -3,11 +3,236 @@
+        xyz : array_like
+           coordinate pairs in cartesian (x, y, z) to query
+        r: distance in degrees within which neighbors are returned
+            r can be a single value , or an array of values of shape x.shape[:-1] if different radii are desired for each point.
```
@philipc2 I haven't actually seen the documentation for having multiple radii, are we sure this is something that works? I sent in an array of points and a matching array of radii but it didn't accept that input.

philipc2 replied:
Good catch, I think I may have copied this docstring section for `sklearn` directly without thinking about it. I'm not sure if it works. I think it would be OK to remove that part from both the KD and Ball Tree and we can investigate it in the future.

aaronzedwick commented:
@philipc2 I think I updated the user api incorrectly causing the docs to fail, anything  you see that would be causing the problem? I just added what you did from the balltree, but using the kdtree instead.

philipc2 commented on docs/user_api/index.rst:
```diff
@@ -175,6 +176,10 @@ Attributes
    Grid.Mesh2_edge_nodes
    Grid.Mesh2_face_edges
    Grid.antimeridian_face_indices
+   Grid.corner_node_balltree
+   Grid.center_node_balltree
+   Grid.corner_node_kdtree
+   Grid.center_node_kdtree
```
@aaronzedwick 

I think these are causing your issue, we don't actually have these attributes.

philipc2 commented on docs/user_api/index.rst:
```diff
@@ -217,13 +222,22 @@ BallTree Data Structure
 
    grid.BallTree
 
+KDTree Data Structure
+-----------------------
```
```suggestion
KDTree Data Structure
------------------------
```

philipc2 reviewed (approved):
Nice work! Please run the pre-commit though!
